File: circular.kt
// Define the Node class
data class Node(var capacity: Int) {
    var next: Node? = null
}

// Define the CircularLinkedList class
class CircularLinkedList {
    private var head: Node? = null

    // Insert at the beginning
    fun insertAtBeginning(capacity: Int) {
        val newNode = Node(capacity)
        if (head == null) {
            head = newNode
            newNode.next = head
        } else {
            var current = head
            while (current?.next != head) {
                current = current?.next
            }
            newNode.next = head
            head = newNode
            current?.next = head
        }
    }

    // Insert at the end
    fun insertAtEnd(capacity: Int) {
        val newNode = Node(capacity)
        if (head == null) {
            head = newNode
            newNode.next = head
        } else {
            var current = head
            while (current?.next != head) {
                current = current?.next
            }
            current?.next = newNode
            newNode.next = head
        }
    }

    // Create a selected node (not added to the list)
    fun createSelectedNode(capacity: Int): Node {
        return Node(capacity)
    }

    // Delete from the beginning
    fun deleteFromBeginning() {
        if (head == null) return
        if (head?.next == head) {  // Only one node in the list
            head = null
        } else {
            var current = head
            while (current?.next != head) {
                current = current?.next
            }
            head = head?.next
            current?.next = head
        }
    }

    // Delete from the end
    fun deleteFromEnd() {
        if (head == null) return
        if (head?.next == head) {  // Only one node in the list
            head = null
        } else {
            var current = head
            var prev: Node? = null
            while (current?.next != head) {
                prev = current
                current = current?.next
            }
            prev?.next = head
        }
    }

    // Print the total capacity
    fun printTotalCapacity(): Int {
        var totalCapacity = 0
        var current = head
        if (head != null) {
            do {
                totalCapacity += current?.capacity ?: 0
                current = current?.next
            } while (current != head)
        }
        return totalCapacity
    }

    // Display the list for verification
    fun display() {
        if (head == null) {
            println("List is empty")
            return
        }
        var current = head
        do {
            print("${current?.capacity} -> ")
            current = current?.next
        } while (current != head)
        println("... (circular)")
    }
}

// Usage Example
fun main() {
    val cll = CircularLinkedList()

    // Insert nodes at the beginning and end
    cll.insertAtBeginning(10)
    cll.insertAtBeginning(20)
    cll.insertAtEnd(30)

    // Create a selected node (not added to list)
    val selectedNode = cll.createSelectedNode(40)
    println("Created selected node with capacity: ${selectedNode.capacity}")

    // Display the list
    println("Initial List:")
    cll.display()

    // Delete from beginning
    cll.deleteFromBeginning()
    println("\nAfter deleting from beginning:")
    cll.display()

    // Delete from end
    cll.deleteFromEnd()
    println("\nAfter deleting from end:")
    cll.display()

    // Print total capacity
    val totalCapacity = cll.printTotalCapacity()
    println("\nTotal Capacity: $totalCapacity")
}

File: circular.c
#include <stdio.h>
#include <stdlib.h>

// Define the Node structure
struct Node {
    int capacity;
    struct Node* next;
};

// Insert at the beginning of the circular linked list
void insertAtBeginning(struct Node** head, int capacity) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->capacity = capacity;
    
    if (*head == NULL) {
        *head = newNode;
        newNode->next = *head;
    } else {
        struct Node* temp = *head;
        while (temp->next != *head) {
            temp = temp->next;
        }
        newNode->next = *head;
        temp->next = newNode;
        *head = newNode;
    }
}

// Insert at the end of the circular linked list
void insertAtEnd(struct Node** head, int capacity) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->capacity = capacity;

    if (*head == NULL) {
        *head = newNode;
        newNode->next = *head;
    } else {
        struct Node* temp = *head;
        while (temp->next != *head) {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->next = *head;
    }
}

// Create a selected node (not part of the list)
struct Node* createSelectedNode(int capacity) {
    struct Node* selectedNode = (struct Node*)malloc(sizeof(struct Node));
    selectedNode->capacity = capacity;
    selectedNode->next = NULL;
    return selectedNode;
}

// Delete from the beginning of the circular linked list
void deleteFromBeginning(struct Node** head) {
    if (*head == NULL) return;

    struct Node* temp = *head;
    struct Node* last = *head;

    while (last->next != *head) {
        last = last->next;
    }

    if (*head == last) { // Only one node in the list
        free(*head);
        *head = NULL;
    } else {
        *head = (*head)->next;
        last->next = *head;
        free(temp);
    }
}

// Delete from the end of the circular linked list
void deleteFromEnd(struct Node** head) {
    if (*head == NULL) return;

    struct Node* temp = *head;
    struct Node* prev = NULL;

    while (temp->next != *head) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == *head) { // Only one node in the list
        free(*head);
        *head = NULL;
    } else {
        prev->next = *head;
        free(temp);
    }
}

// Print the total capacity of all nodes
int printTotalCapacity(struct Node* head) {
    int totalCapacity = 0;
    if (head == NULL) return totalCapacity;

    struct Node* temp = head;
    do {
        totalCapacity += temp->capacity;
        temp = temp->next;
    } while (temp != head);

    return totalCapacity;
}

// Display the circular linked list
void display(struct Node* head) {
    if (head == NULL) {
        printf("List is empty\n");
        return;
    }

    struct Node* temp = head;
    do {
        printf("%d -> ", temp->capacity);
        temp = temp->next;
    } while (temp != head);
    printf("(circular)\n");
}

// Main function to demonstrate functionality
int main() {
    struct Node* head = NULL;

    // Insert nodes
    insertAtBeginning(&head, 10);
    insertAtBeginning(&head, 20);
    insertAtEnd(&head, 30);

    // Create a selected node (not added to list)
    struct Node* selectedNode = createSelectedNode(40);
    printf("Selected Node Capacity: %d\n", selectedNode->capacity);

    // Display the list
    printf("Initial List:\n");
    display(head);

    // Delete from beginning
    deleteFromBeginning(&head);
    printf("\nAfter deleting from beginning:\n");
    display(head);

    // Delete from end
    deleteFromEnd(&head);
    printf("\nAfter deleting from end:\n");
    display(head);

    // Print total capacity
    int totalCapacity = printTotalCapacity(head);
    printf("\nTotal Capacity: %d\n", totalCapacity);

    // Free the selected node memory
    free(selectedNode);

    return 0;
}